composer/utils/fx_utils.py

```diff
@@ -123,16 +159,111 @@ def detect_residual_pattern(gm: GraphModule):
+            their common ancestors removed.
+    """
+    lhsAncestors = _get_ancestors(nodeLHS)
+    rhsAncestors = _get_ancestors(nodeRHS)
+
+    # Iterate from back and eliminate common nodes
+    while lhsAncestors and rhsAncestors and lhsAncestors[-1] == rhsAncestors[-1]:  # type: ignore
+        del lhsAncestors[-1]
+        del rhsAncestors[-1]
+    lhsAncestors.reverse()
+    rhsAncestors.reverse()
+    return lhsAncestors, rhsAncestors
+
+
+def _attach_tag(nodes: List[Node], tag: str):
+    """Attach tag to the given nodes for the splitter."""
+    for node in nodes:
+        node.tag = tag  # type: ignore[attr-defined]
+
+
+def _tag_nodes(gm: GraphModule) -> Tuple[List[str], int]:
+    """Tag nodes for splitting."""
+    # all nodes that are not a part of the residual blocks are tagged with "main_0".
+    # a tag is required for all nodes by split_by_tags
+    # Also an earlier tag can be repeated for later nodes.
+    count = 0
+    all_tags = []
+    # In this pass over all nodes, we just tag them
+    for node in gm.graph.nodes:
+        default_tag = f'mainN_{count}'
+        node.tag = default_tag
+        if default_tag not in all_tags:
+            all_tags.append(default_tag)
+        if node.op == 'call_function' and node.target in [torch.add, operator.add]:
+            assert len(node.all_input_nodes) == 2
+            node0, node1 = node.all_input_nodes[0], node.all_input_nodes[1]
+            lhs_nodes, rhs_nodes = _get_residual_path(node0, node1)
+            if lhs_nodes or rhs_nodes:
+                if len(lhs_nodes):
+                    _attach_tag(lhs_nodes, f'non_res_{count}')
+                    all_tags.append(f'non_res_{count}')
+                if len(rhs_nodes):
+                    _attach_tag(rhs_nodes, f'residual_{count}')
+                    all_tags.append(f'residual_{count}')
+                add_tag = f'addN_{count}'
+                if add_tag not in all_tags:
+                    all_tags.append(add_tag)
+                node.tag = add_tag
+                count += 1
+    return all_tags, count
+
+
+def apply_stochastic_residual(gm: GraphModule, drop_rate: float = 0.2) -> Tuple[GraphModule, int]:
+    """Detect and replace residual pattern with their stochastic equivalent.
+
+    Arguments:
+        gm (GraphModule): The source FX-traced graph. It can be the whole model symbolically traced.
+
+    Returns:
+        GraphModule: Modified GraphModule that has stochastic residual connections.
     """
-    raise NotImplementedError('replace_residual_with_stochastic is currently not implemented.')
+    assert isinstance(gm, GraphModule), 'Input to apply_stochastic_residual should be an instance of GraphModule'
+    all_tags, count = _tag_nodes(gm)
+    split_gm = split_by_tags(gm, all_tags)
+    pattern = re.compile(r'non_res_(\d+)|residual_(\d+)')
+    for node in split_gm.graph.nodes:
+        if node.op == 'call_module':
+            matches = pattern.match(node.target)
+            if matches:
```

Comment:
why could torch.fx's `replace_pattern` not work here? 

the readability of this code is a bit difficult, as the stochatic depth specific bits are mixed in with what seems like more general torch.fx graph manipulation methods. Is it possible to refactor this into helper functions for bette readability and also encapsulation (e.g. the regex pattern and getting the lsh_submod/rhs_submod should be encapsulated)? For example..
```
for node in split_gm.graph.nodes:
    if node.op != 'call_module':
        continue

    module = _get_residual_module(node)
    if module:
        main, residual = module
        new_module = BlockStochasticModule(main, residual, drop_rate)
        split_gm.add_module('...', new_module)

        _replace_call_module(split_gm, original=node, replace=new_module)
```   
or something like that?
     


```

Reply:
`replace_pattern` doesn't work here for a number of reasons. The main challenging part in this was the dynamic conditional and is the reason why we need to split into subgraphs of graphModule so that we can insert conditional calls to them. 

- The pattern we want to replace has dynamic conditional. For `replace_pattern`, pattern + replacement both should be be FX-traceable. 
- replace_pattern replaces with new modules/functions so doesn't preserve weights. 

Will try to refactor but most of the refactored functions will not be reusable as they are very specific to stochastic depth. We will have our own `replace_pattern` implementation that takes care of optimizer state.

Reply:
refactored as suggested.